### PR TITLE
Query compatibility with Oracle

### PR DESCRIPTION
**Pull request by vinumaddumage**

### CODE REVIEW

File: components/org.wso2.carbon.identity.organization.management.role.management.service/src/main/java/org/wso2/carbon/identity/organization/management/role/management/service/constant/SQLConstants.java [hunk 1]
```diff
@@ -34,31 +34,50 @@ public class SQLConstants {
...
             "VALUES ";
 
-    public static final String CHECK_ROLE_NAME_EXISTS = "SELECT COUNT(1) FROM UM_ORG_ROLE WHERE UM_ROLE_NAME=:" +
-            SQLPlaceholders.DB_SCHEMA_COLUMN_NAME_UM_ROLE_NAME + "; AND UM_ORG_ID=:" +
-            SQLPlaceholders.DB_SCHEMA_COLUMN_NAME_UM_ORG_ID + ";";
-
-    public static final String CHECK_ROLE_EXISTS = "SELECT COUNT(1) FROM UM_ORG_ROLE WHERE UM_ROLE_ID=:" +
-            SQLPlaceholders.DB_SCHEMA_COLUMN_NAME_UM_ROLE_ID + "; AND UM_ORG_ID=:" +
-            SQLPlaceholders.DB_SCHEMA_COLUMN_NAME_UM_ORG_ID + ";";
-
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.organization.management.role.management.service/src/main/java/org/wso2/carbon/identity/organization/management/role/management/service/constant/SQLConstants.java:44]: why have we remove this here and readded it at https://github.com/wso2-extensions/identity-organization-management/pull/65/files#diff-2515926efc8856445f792744d91bf51348b07a65041def5ce054fd51ba2b5bf6R92-R99

File: components/org.wso2.carbon.identity.organization.management.role.management.service/src/main/java/org/wso2/carbon/identity/organization/management/role/management/service/constant/SQLConstants.java [hunk 2]
```diff
@@ -34,31 +34,50 @@ public class SQLConstants {
...
     public static final String ADD_ROLE_GROUP_MAPPING_INSERT_VALUES = "(:" +
             SQLPlaceholders.DB_SCHEMA_COLUMN_NAME_UM_GROUP_ID + "%1$d;,:" +
             SQLPlaceholders.DB_SCHEMA_COLUMN_NAME_UM_ROLE_ID + "%1$d;)";
 
+    public static final String ADD_ROLE_GROUP_MAPPING_ORACLE = "INSERT INTO UM_ORG_ROLE_GROUP (UM_GROUP_ID, " +
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.organization.management.role.management.service/src/main/java/org/wso2/carbon/identity/organization/management/role/management/service/constant/SQLConstants.java:41]: shall we have this query right after `ADD_ROLE_GROUP_MAPPING`?

File: components/org.wso2.carbon.identity.organization.management.role.management.service/src/main/java/org/wso2/carbon/identity/organization/management/role/management/service/constant/SQLConstants.java [hunk 3]
```diff
@@ -34,31 +34,50 @@ public class SQLConstants {
...
+
+    public static final String ADD_ROLE_GROUP_MAPPING_TAIL_ORACLE = ") SELECT * FROM orgRoleGroups";
+
     public static final String ADD_ROLE_USER_MAPPING = "INSERT INTO UM_ORG_ROLE_USER (UM_USER_ID, UM_ROLE_ID) VALUES ";
 
     public static final String ADD_ROLE_USER_MAPPING_INSERT_VALUES = "(:" +
             SQLPlaceholders.DB_SCHEMA_COLUMN_NAME_UM_USER_ID + "%1$d;,:" +
             SQLPlaceholders.DB_SCHEMA_COLUMN_NAME_UM_ROLE_ID + "%1$d;)";
 
+    public static final String ADD_ROLE_USER_MAPPING_ORACLE = "INSERT INTO UM_ORG_ROLE_USER (UM_USER_ID, UM_ROLE_ID) " +
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.organization.management.role.management.service/src/main/java/org/wso2/carbon/identity/organization/management/role/management/service/constant/SQLConstants.java:56]: let's have this query right after ADD_ROLE_USER_MAPPING

File: components/org.wso2.carbon.identity.organization.management.role.management.service/src/main/java/org/wso2/carbon/identity/organization/management/role/management/service/dao/RoleManagementDAOImpl.java [hunk 4]
```diff
@@ -156,39 +177,70 @@ public void createRole(String organizationId, Role role) throws OrganizationMana
...
-                    int tenantID = getTenantId();
-                    List<String> nonExistingPermissions = getNonExistingPermissions(role.getPermissions(), tenantID,
-                            role.getId());
-                    addNonExistingPermissions(nonExistingPermissions, role.getId(), tenantID);
-                    List<String> permissionList = getPermissionIds(role.getPermissions(),
-                            tenantID).stream().map(Object::toString).collect(Collectors.toList());
-                    String query = buildQueryForInsertAndRemoveValues(permissionList.size(),
-                            ADD_ROLE_PERMISSION_MAPPING, ADD_ROLE_PERMISSION_MAPPING_INSERT_VALUES, COMMA_SEPARATOR);
-                    assignRoleAttributes(permissionList, role.getId(), query, PERMISSIONS);
+                if (driverName.contains(ORACLE)) {
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.organization.management.role.management.service/src/main/java/org/wso2/carbon/identity/organization/management/role/management/service/dao/RoleManagementDAOImpl.java:188]: will require a similar approach to https://github.com/wso2/carbon-identity-framework/blob/c750169c4818afed19155b8a0fbd9549b63ae115/components/user-functionality-mgt/org.wso2.carbon.identity.user.functionality.mgt/src/main/java/org/wso2/carbon/identity/user/functionality/mgt/dao/impl/UserFunctionalityPropertyDAOImpl.java#L67-L72

File: components/org.wso2.carbon.identity.organization.management.role.management.service/src/main/java/org/wso2/carbon/identity/organization/management/role/management/service/dao/RoleManagementDAOImpl.java [hunk 5]
```diff
@@ -156,39 +177,70 @@ public void createRole(String organizationId, Role role) throws OrganizationMana
...
+                        List<String> nonExistingPermissions = getNonExistingPermissions(role.getPermissions(), tenantID,
+                                role.getId());
+                        addNonExistingPermissions(nonExistingPermissions, role.getId(), tenantID);
+                        List<String> permissionList = getPermissionIds(role.getPermissions(),
+                                tenantID).stream().map(Object::toString).collect(Collectors.toList());
+                        String query = buildQueryForInsertAndRemoveValues(permissionList.size(),
+                                ADD_ROLE_PERMISSION_MAPPING, ADD_ROLE_PERMISSION_MAPPING_INSERT_VALUES,
+                                COMMA_SEPARATOR);
+                        assignRoleAttributes(permissionList, role.getId(), query, PERMISSIONS);
+                    }
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.organization.management.role.management.service/src/main/java/org/wso2/carbon/identity/organization/management/role/management/service/dao/RoleManagementDAOImpl.java:239]: only the queries change here right? therefore better to have private methods and pass the query as a param to remove any unnecessary code duplications. check all applicable places.

File: components/org.wso2.carbon.identity.organization.management.role.management.service/src/main/java/org/wso2/carbon/identity/organization/management/role/management/service/dao/RoleManagementDAOImpl.java [hunk 6]
```diff
@@ -156,39 +177,70 @@ public void createRole(String organizationId, Role role) throws OrganizationMana
...
                 }
                 return null;
             });
-        } catch (TransactionException e) {
+        } catch (TransactionException | DataAccessException e) {
```
Comment (dewniMW) [inline components/org.wso2.carbon.identity.organization.management.role.management.service/src/main/java/org/wso2/carbon/identity/organization/management/role/management/service/dao/RoleManagementDAOImpl.java:243]: better to handle DataAccessException separately.
ex: https://github.com/wso2/carbon-identity-framework/blob/c750169c4818afed19155b8a0fbd9549b63ae115/components/configuration-mgt/org.wso2.carbon.identity.configuration.mgt.core/src/main/java/org/wso2/carbon/identity/configuration/mgt/core/dao/impl/ConfigurationDAOImpl.java#L725-L727

**Review by AnuradhaSK — APPROVED** [review]

File: components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/util/Utils.java [hunk 7]
```diff
@@ -88,6 +91,22 @@ public static NamedJdbcTemplate getNewTemplate() {
         return new NamedJdbcTemplate(UmPersistenceManager.getInstance().getDataSource());
     }
 
+    /**
+     * Check whether the string, "oracle", contains in the driver name or db product name.
+     *
+     * @return true if the database type matches the driver type, false otherwise.
+     * @throws OrganizationManagementServerException If error occurred while checking the DB metadata.
+     */
+    public static boolean isOracleDB() throws OrganizationManagementServerException {
```
Comment (piraveena) [inline components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/util/Utils.java:100]: Add new line after method signature

Comment (vinumaddumage) [reply components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/util/Utils.java:100]: Has been addressed as seen in the latest code base (note that the module was moved to a new repo): https://github.com/wso2-extensions/identity-organization-management-core/blob/85c42c315a34f64a45ef48e741f91c18212a17ff/components/org.wso2.carbon.identity.organization.management.service/src/main/java/org/wso2/carbon/identity/organization/management/service/util/Utils.java#L124-L127

**Pull request merged**